src/sync/semaphore.rs

```diff
@@ -555,6 +557,279 @@ mod if_alloc {
+                semaphore_state.try_acquire(&mut mut_self.wait_node, cx)
+            };
+
+            match poll_res {
+                Poll::Pending => {
+                    mut_self.semaphore.replace(semaphore);
+                    Poll::Pending
+                }
+                Poll::Ready(()) => {
+                    let to_release = match mut_self.auto_release {
+                        true => mut_self.wait_node.required_permits,
+                        false => 0,
+                    };
+                    Poll::Ready(GenericSharedSemaphoreReleaser::<MutexType> {
+                        semaphore,
+                        permits: to_release,
+                    })
+                }
+            }
+        }
+    }
+
+    impl<MutexType: RawMutex> FusedFuture
+        for GenericSharedSemaphoreAcquireFuture<MutexType>
+    {
+        fn is_terminated(&self) -> bool {
+            self.semaphore.is_none()
+        }
+    }
+
+    impl<MutexType: RawMutex> Drop
+        for GenericSharedSemaphoreAcquireFuture<MutexType>
+    {
+        fn drop(&mut self) {
+            // If this GenericSharedSemaphoreAcquireFuture has been polled and it was added to the
+            // wait queue at the semaphore, it must be removed before dropping.
+            // Otherwise the semaphore would access invalid memory.
+            if let Some(semaphore) = self.semaphore.take() {
+                let mut semaphore_state = semaphore.state.lock();
+                // Analysis: Does the number of permits play a role here?
+                // The future was notified because there was a certain amount of permits
+                // available.
+                // Removing the waiter will wake up as many tasks as there are permits
+                // available inside the Semaphore now. If this is bigger than the
+                // amount of permits required for this task, then additional new
+                // tasks might get woken. However that isn't bad, since
+                // those tasks should get into the wait state anyway.
+                semaphore_state.remove_waiter(&mut self.wait_node);
+            }
+        }
+    }
+
+    /// A futures-aware shared semaphore.
+    pub struct GenericSharedSemaphore<MutexType: RawMutex> {
+        state: Arc<LockApiMutex<MutexType, SemaphoreState>>,
+    }
+
+    impl<MutexType: RawMutex> PartialEq for GenericSharedSemaphore<MutexType> {
+        fn eq(&self, other: &Self) -> bool {
+            Arc::ptr_eq(&self.state, &other.state)
+        }
+    }
+
+    impl<MutexType: RawMutex> Eq for GenericSharedSemaphore<MutexType> {}
+
+    impl<MutexType: RawMutex> Clone for GenericSharedSemaphore<MutexType> {
+        fn clone(&self) -> Self {
+            Self {
+                state: self.state.clone(),
+            }
+        }
+    }
+
+    // It is safe to send semaphores between threads, as long as they are not used and
+    // thereby borrowed
+    unsafe impl<MutexType: RawMutex + Send> Send
```

Comment:
I think this might also need `RawMutex + Sync` - since now the `Semaphore` could be moved while a `Future` is still referencing it from another thread.

For the non shared semaphore this wasn't an issue, since the value could not be moved with active references on it.

Reply:
Gotcha. So if you have something that refers to the `Semaphore` from the same thread but then the `Semaphore` is sent to another thread then this indirectly creates a reference across a thread, which would be invalid unless `MutexType` is `Sync`. That's tricky.